cpp/src/round/round.cu

```diff
@@ -47,102 +52,192 @@ T __device__ generic_abs(T value)
+    return integer_part + generic_round(fractional_part * n) / n;
+  }
+
+  template <typename U = T, typename std::enable_if_t<std::is_integral<U>::value>* = nullptr>
+  __device__ U operator()(U e)
+  {
+    assert(false);  // Should never get here. Just for compilation
+    return U{};
+  }
+};
+
+template <typename T>
+struct half_up_negative {
+  T n;
+  template <typename U = T, typename std::enable_if_t<cudf::is_floating_point<U>()>* = nullptr>
+  __device__ U operator()(U e)
+  {
+    return generic_round(e / n) * n;
+  }
+
+  template <typename U = T, typename std::enable_if_t<std::is_integral<U>::value>* = nullptr>
+  __device__ U operator()(U e)
+  {
+    auto const down = (e / n) * n;  // result from rounding down
+    return down + generic_sign(e) * (generic_abs(e - down) >= n / 2 ? n : 0);
+  }
+};
+
+template <typename T>
+struct half_even_zero {
+  T n;  // unused in the decimal_places = 0 case
+  template <typename U = T, typename std::enable_if_t<cudf::is_floating_point<U>()>* = nullptr>
+  __device__ U operator()(U e)
   {
-    auto result = cudf::make_fixed_width_column(input.type(),  //
-                                                input.size(),
-                                                copy_bitmask(input, stream, mr),
-                                                input.null_count(),
-                                                stream,
-                                                mr);
-
-    auto out_view = result->mutable_view();
-    T const n     = std::pow(10, std::abs(decimal_places));
-
-    if (decimal_places == 0)
-      thrust::transform(rmm::exec_policy(stream)->on(stream),
-                        input.begin<T>(),
-                        input.end<T>(),
-                        out_view.begin<T>(),
-                        [] __device__(T e) -> T { return generic_round(e); });
-    else if (decimal_places > 0)
-      thrust::transform(rmm::exec_policy(stream)->on(stream),
-                        input.begin<T>(),
-                        input.end<T>(),
-                        out_view.begin<T>(),
-                        [n] __device__(T e) -> T {
-                          T integer_part;
-                          T const fractional_part = generic_modf(e, &integer_part);
-                          return integer_part + generic_round(fractional_part * n) / n;
-                        });
-    else  // decimal_places < 0
-      thrust::transform(rmm::exec_policy(stream)->on(stream),
-                        input.begin<T>(),
-                        input.end<T>(),
-                        out_view.begin<T>(),
-                        [n] __device__(T e) -> T { return generic_round(e / n) * n; });
-
-    return result;
+    return generic_round_half_even(e);
+  }
+
+  template <typename U = T, typename std::enable_if_t<std::is_integral<U>::value>* = nullptr>
+  __device__ U operator()(U e)
+  {
+    assert(false);  // Should never get here. Just for compilation
+    return U{};
+  }
```

Comment:
Should be possible to eliminate this?

Reply:
Wait until PR 3 and then see if something is really needed.

Reply:
That is what I was thinking, if Jake comes up with something while not sleeping - we can clean this up in PR 3